core/src/main/scala/io/finch/Compile.scala

```diff
@@ -79,72 +78,57 @@ object ToService {
     rep
   }
 
-  implicit val hnilTS: ToService[HNil, HNil] = new ToService[HNil, HNil] {
-    def apply(es: HNil, opts: Options, ctx: Context): Service[Request, Response] =
-      new Service[Request, Response] {
-        def apply(req: Request): Future[Response] = {
-          val rep = Response()
+  implicit def hnilTS[F[_]](implicit F: Applicative[F]): Compile[F, HNil, HNil] = new Compile[F, HNil, HNil] {
+    def apply(es: HNil, opts: Options, ctx: Context): Endpoint.Compiled[F] =
+      new Endpoint.Compiled[F]((req: Request) => {
+        val rep = Response()
 
-          if (ctx.wouldAllow.nonEmpty && opts.enableMethodNotAllowed) {
-            rep.status = Status.MethodNotAllowed
-            rep.allow = ctx.wouldAllow
-          } else {
-            rep.status = Status.NotFound
-          }
-
-          Future.value(conformHttp(rep, req.version, opts))
+        if (ctx.wouldAllow.nonEmpty && opts.enableMethodNotAllowed) {
+          rep.status = Status.MethodNotAllowed
+          rep.allow = ctx.wouldAllow
+        } else {
+          rep.status = Status.NotFound
         }
-      }
+
+        F.pure(Trace.empty -> Right(conformHttp(rep, req.version, opts)))
+      })
   }
 
   type IsNegotiable[C] = OrElse[C <:< Coproduct, DummyImplicit]
 
   implicit def hlistTS[F[_], A, EH <: Endpoint[F, A], ET <: HList, CTH, CTT <: HList](implicit
     ntrA: ToResponse.Negotiable[F, A, CTH],
     ntrE: ToResponse.Negotiable[F, Exception, CTH],
-    F: Effect[F],
-    tsT: ToService[ET, CTT],
+    F: MonadError[F, Throwable],
+    tsT: Compile[F, ET, CTT],
     isNegotiable: IsNegotiable[CTH]
-  ): ToService[Endpoint[F, A] :: ET, CTH :: CTT] = new ToService[Endpoint[F, A] :: ET, CTH :: CTT] {
-    def apply(es: Endpoint[F, A] :: ET, opts: Options, ctx: Context): Service[Request, Response] =
-      new Service[Request, Response] {
-        private[this] val handler =
-          if (opts.enableUnsupportedMediaType) respond415.orElse(respond400) else respond400
-
-        private[this] val negotiateContent = isNegotiable.fold(_ => true, _ => false)
-        private[this] val underlying = es.head.handle(handler)
-
-        def apply(req: Request): Future[Response] = underlying(Input.fromRequest(req)) match {
+  ): Compile[F, Endpoint[F, A] :: ET, CTH :: CTT] = new Compile[F, Endpoint[F, A] :: ET, CTH :: CTT] {
+    def apply(es: Endpoint[F, A] :: ET, opts: Options, ctx: Context): Endpoint.Compiled[F] = {
+      val handler = if (opts.enableUnsupportedMediaType) respond415.orElse(respond400) else respond400
+      val negotiateContent = isNegotiable.fold(_ => true, _ => false)
+      val underlying = es.head.handle(handler)
+
+      new Endpoint.Compiled[F]((req: Request) => {
+        underlying(Input.fromRequest(req)) match {
           case EndpointResult.Matched(rem, trc, out) if rem.route.isEmpty =>
 
             Trace.captureIfNeeded(trc)
```

Comment:
I don't think this should be happening here as we're not dealing with Twitter Futures yet. Let's move that under `toService`.

Reply:
👍